docs/src/apimanual.md

```diff
@@ -911,27 +946,46 @@ Avoid storing extra copies of the problem when possible. This means that solver
+MOI defines a very general interface, with multiple possible ways to describe
+the same constraint. This is considered a feature, not a bug. MOI is designed to
+make it possible to experiment with alternative representations of an
+optimization problem at both the solving and modeling level. The following bullet
+points show examples of how JuMP constraints are translated into MOI
+function-set pairs:
  - `@constraint(m, 2x + y <= 10)` becomes `ScalarAffineFunction`-in-`LessThan`;
  - `@constraint(m, 2x + y >= 10)` becomes `ScalarAffineFunction`-in-`GreaterThan`;
  - `@constraint(m, 2x + y == 10)` becomes `ScalarAffineFunction`-in-`EqualTo`;
  - `@constraint(m, 0 <= 2x + y <= 10)` becomes `ScalarAffineFunction`-in-`Interval`;
- - `@constraint(m, 2x + y in ArbitrarySet())` becomes `ScalarAffineFunction`-in-`ArbitrarySet`.
+ - `@constraint(m, 2x + y in ArbitrarySet())` becomes
+   `ScalarAffineFunction`-in-`ArbitrarySet`.
 
 Variable bounds are handled in a similar fashion:
  - `@variable(m, x <= 1)` becomes `SingleVariable`-in-`LessThan`;
  - `@variable(m, x >= 1)` becomes `SingleVariable`-in-`GreaterThan`.
 
-One notable difference is that a variable with an upper and lower bound is translated into two constraints, rather than an interval. i.e.:
- - `@variable(m, 0 <= x <= 1)` becomes `SingleVariable`-in-`LessThan` *and* `SingleVariable`-in-`GreaterThan`.
-
-Therefore, if a solver wrapper does not support `ScalarAffineFunction`-in-`LessThan` constraints, users will not be able to write: `@constraint(m, 2x + y <= 10)` in JuMP. With this in mind, developers should support all the constraint types that they want to be usable from JuMP. That said, from the perspective of JuMP, solvers can safely choose to not support the following constraints:
-
-- `ScalarAffineFunction` in `GreaterThan`, `LessThan`, or `EqualTo` with a nonzero constant in the function. Constants in the affine function should instead be moved into the parameters of the corresponding sets.
-
-- `ScalarAffineFunction` in `Nonnegative`, `Nonpositive` or `Zeros`. Alternative constraints are available by using a `VectorAffineFunction` with one output row or `ScalarAffineFunction` with `GreaterThan`, `LessThan`, or `EqualTo`.
-
-- Two `SingleVariable`-in-`LessThan` constraints applied to the same variable (similarly with `GreaterThan`). These should be interpreted as variable bounds, and each variable naturally has at most one upper or lower bound.
+One notable difference is that a variable with an upper and lower bound is
+translated into two constraints, rather than an interval. i.e.:
+ - `@variable(m, 0 <= x <= 1)` becomes `SingleVariable`-in-`LessThan` *and*
+   `SingleVariable`-in-`GreaterThan`.
+
+Therefore, if a solver wrapper does not support
+`ScalarAffineFunction`-in-`LessThan` constraints, users will not be able to
```

Comment:
This still need to be updated. If the model does support `ScalarAffineFunction`-in-`LessThan` but supports `VectorAffineFunction`-in-`Nonpositives` or `VectorAffineFunction`-in-`Nonnegatives` or ... it should still be ok